includes/class-wc-payments-order-service.php

```diff
@@ -1477,42 +1479,78 @@ public function add_note_and_metadata_for_created_refund( WC_Order $order, WC_Or
 		// Delete the refund if it exists.
 		if ( $wc_refund ) {
 			$wc_refund->delete();
 		}
 
-		$note = sprintf(
-			WC_Payments_Utils::esc_interpolated_html(
-				/* translators: %1: the refund amount, %2: WooPayments, %3: ID of the refund */
-				__( 'A refund of %1$s was <strong>%4$s</strong> using %2$s (<code>%3$s</code>).', 'woocommerce-payments' ),
-				[
-					'strong' => '<strong>',
-					'code'   => '<code>',
-				]
-			),
-			WC_Payments_Explicit_Price_Formatter::get_explicit_price(
-				wc_price( WC_Payments_Utils::interpret_stripe_amount( $amount, $currency ), [ 'currency' => strtoupper( $currency ) ] ),
-				$order
-			),
-			'WooPayments',
-			$refund_id,
-			$is_cancelled ? __( 'cancelled', 'woocommerce-payments' ) : __( 'unsuccessful', 'woocommerce-payments' )
+		$formatted_amount = WC_Payments_Explicit_Price_Formatter::get_explicit_price(
+			wc_price( WC_Payments_Utils::interpret_stripe_amount( $amount, $currency ), [ 'currency' => strtoupper( $currency ) ] ),
+			$order
 		);
 
+		if ( $is_cancelled ) {
+			$note = sprintf(
+				WC_Payments_Utils::esc_interpolated_html(
+					/* translators: %1: the refund amount, %2: WooPayments, %3: ID of the refund */
+					__( 'A refund of %1$s was <strong>cancelled</strong> using %2$s (<code>%3$s</code>).', 'woocommerce-payments' ),
+					[
+						'strong' => '<strong>',
+						'code'   => '<code>',
+					]
+				),
+				$formatted_amount,
+				'WooPayments',
+				$refund_id
+			);
+		} else {
+			$failure_message = $failure_reason ? Refund_Failure_Reason::get_failure_message( $failure_reason ) : __( 'Unknown error occurred', 'woocommerce-payments' );
+			$note            = sprintf(
+				WC_Payments_Utils::esc_interpolated_html(
+					/* translators: %1: the refund amount, %2: WooPayments, %3: ID of the refund, %4: failure message */
+					__( 'A refund of %1$s was <strong>unsuccessful</strong> using %2$s (<code>%3$s</code>): %4$s', 'woocommerce-payments' ),
+					[
+						'strong' => '<strong>',
+						'code'   => '<code>',
+					]
+				),
+				$formatted_amount,
+				'WooPayments',
+				$refund_id,
+				$failure_message
+			);
+		}
+
 		if ( $this->order_note_exists( $order, $note ) ) {
 			return;
 		}
 
-		// If order has been fully refunded.
+		Logger::log( $note );
+		$order->add_order_note( $note );
+
+		// Track the refund failure event.
+		Tracker::track_admin( 'wcpay_edit_order_refund_failure', [ 'reason' => $note ] );
+
+		// Handle insufficient balance case.
+		if ( 'insufficient_balance_for_refund' === $failure_reason ) {
+			$account_country = WC_Payments::get_account_service()->get_account_country();
+
+			if ( $this->is_frod_supported( $account_country ) ) {
+				$order->add_order_note( $this->get_frod_support_note( $formatted_amount ) );
+			} else {
+				$order->add_order_note( $this->get_insufficient_balance_note( $formatted_amount ) );
+			}
+		}
```

Comment:
Sounds like this special handling can reuse https://github.com/Automattic/woocommerce-payments/blob/e7058eec53c5a8890a9796e0d9de9abd0936723e/includes/class-wc-payments-order-service.php#L2304

Reply:
100%, I overlooked those are different errors, again 🤦‍♂️ Your suggestion to reuse `handle_insufficient_balance_for_refund` makes a lot of sense. I'm going with it in this commit [713b4e7](https://github.com/Automattic/woocommerce-payments/pull/10677/commits/713b4e7dc607368ad9b6f977d642e8f931e2fe64) and also making a refactor to try to simplify it a bit. Thanks again!